components/supervisor/pkg/ports/ports.go

```diff
@@ -356,15 +362,37 @@ func (pm *Manager) nextState() map[uint32]*managedPort {
 }
 
 func (pm *Manager) autoExpose(ctx context.Context, mp *managedPort, public bool) {
-	err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, public)
+	err := pm.autoExposeWithBackoff(ctx, mp, public)
 	if err != nil {
-		log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+		if err != context.Canceled {
+			log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+		}
 		return
 	}
 	pm.autoExposed[mp.LocalhostPort] = mp.GlobalPort
 	log.WithField("port", *mp).Info("auto-expose port")
 }
 
+func (pm *Manager) autoExposeWithBackoff(ctx context.Context, mp *managedPort, public bool) error {
```

Comment:
This could still block port detection indefinitely, because we're in the hot path of `newState`. 
The context here will only be canceled on supervisor shutdown.

We should introduce a maximum number of retries.
Even better would be to make this auto exposure mechanism async, s.t. it doesn't block other ports. But that's a much bigger change.

Reply:
It only blocks if deadline exceeded. I was assuming that it does not matter which port is being exposed, so if the server/ws-manager fails for one then it will be the same for another port and blocking till the service is available is fine.

Reply:
In my original comment I mixed two concerns:
1. the unlimited number of retries
2. the synchronous implementation of this operation

The former I think is the biggest concern (esp. conjunction with the other one). It's also one that's easily fixed: simply add a bound on the for loop.

The latter I reckon requires more refactoring and is beyond the scope of this PR. If we don't do this though, from a user's perspective ports will look pretty broken still. If this mechanism would not block, we'd at least still detect local port changes and provide feedback on them.

Reply:
I reimplemented it to limit attempts of retries and expose one by one asynchronously. It still can lead to the state though that a port got stuck after max attempts of retries.